File: graph/22.cpp
#include <bits/stdc++.h>

using namespace std;

using Matrix = vector<vector<int>>;

bool isValid(Matrix &G, Matrix &visited, int row, int col) {
    return row >= 0 && row < G.size() && col >= 0 && col < G[0].size() &&
           visited[row][col] == 0 && G[row][col] == 1;
}

void DFS(Matrix &G, Matrix &visited, int row, int col, int &cont) {
    visited[row][col] = 1;
    ++cont;

    vector<pair<int, int>> directions = {{0, -1}, {-1, -1}, {-1, 0}, {-1, 1},
                                         {0, 1},  {1, 1},   {1, 0},  {-1, 1}};

    for (const auto &coor : directions) {
        int newRow = coor.first + row;
        int newCol = coor.second + col;

        if (isValid(G, visited, newRow, newCol)) DFS(G, visited, newRow, newCol, cont);
    }
}

int maxAreaIsland(Matrix &G) {
    Matrix visited(G.size(), vector<int>(G[0].size(), 0));
    int maxArea = 0;

    for (int i = 0; i < G.size(); ++i) {
        for (int j = 0; j < G[i].size(); ++j) {
            int cont = 0;
            if (visited[i][j] == 0 && G[i][j] == 1) {
                DFS(G, visited, i, j, cont);
                cout << cont << "\n";
                if (cont > maxArea) maxArea = cont;
            }
        }
    }
    return maxArea;
}

int main() {
    Matrix G(5);
    G[0] = {1, 0, 0, 0, 1, 0, 0};
    G[1] = {0, 1, 0, 0, 1, 1, 0};
    G[2] = {1, 1, 0, 0, 0, 0, 0};
    G[3] = {1, 0, 0, 1, 1, 0, 0};
    G[4] = {1, 0, 0, 1, 0, 1, 1};

    cout << maxAreaIsland(G);

    return 0;
}

File: graph/22.py
def is_valid(board, visited, row, col):
    return (
        row >= 0
        and row < len(board)
        and col >= 0
        and col < len(board[0])
        and visited[row][col] == 0
        and board[row][col] == 1
    )


def max_area_island(board):
    visited = [[0] * len(row) for row in board]
    cont = 0
    max_area = 0
    directions = [(0, -1), (-1, -1), (-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (-1, 1)]

    def DFS(row, col, cont):
        visited[row][col] = 1
        cont[0] += 1

        for item in directions:
            newRow = row + item[0]
            newCol = col + item[1]

            if is_valid(board, visited, newRow, newCol):
                DFS(newRow, newCol, cont)

    for row in range(len(board)):
        for col in range(len(board[row])):
            if visited[row][col] == 0 and board[row][col] == 1:
                cont = [0]
                DFS(row, col, cont)
                if cont[0] > max_area:
                    max_area = cont[0]

    return max_area


board = [
    [1, 0, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 1, 0],
    [1, 1, 0, 0, 0, 0, 0],
    [1, 0, 0, 1, 1, 0, 0],
    [1, 0, 0, 1, 0, 1, 1],
]

print(max_area_island(board))
